### ISSUE TITLE
Nesting depth rule (statement-max-nesting-depth)?

### ISSUE THREAD

**Issue by ahmedelgabri** [stylelint/stylelint#291]
Any plan to add a rule for how many level of nesting is allowed? for example `scss-lint` has a `NestingDepth` rule.

This is one of my main issues with preprocessors when working with teams of different CSS skill levels. Some of them just abuse the nesting feature.

**Comment by davidtheclark** [comment]
I agree that's a good rule in `scss-lint`. My only hesitation about adding it here is that until nesting is *standard* CSS, the rule you're proposing would be a *plugin-specific* warning. If we go down that route, we might dig ourselves into some holes.

It might be worth asking the authors of the nesting plugin you use if they're interested in implementing some kind nesting-level-limit warning.

I don't know. What do @stylelint/owners think?

**Comment by jeddy3** [comment]
> It might be worth asking the authors of the nesting plugin you use if they're interested in implementing some kind nesting-level-limit warning.

That's a great idea!

**Comment by ahmedelgabri** [comment]
> It might be worth asking the authors of the nesting plugin you use if they're interested in implementing some kind nesting-level-limit warning.

I agree, this seems like a good idea.

**Comment by ai** [comment]
@davidtheclark I afraid that in near future we will have more and more non-spec linting suggestion. Maybe we can do some stylelint-postcss-plugins pack with this non-standard linters?

**Comment by davidtheclark** [comment]
I create issue #295 for building and documenting custom rules. We can use this one as a prototype to build out the functionality. Sound good?

**Comment by jeddy3** [comment]
> We can use this one as a prototype to build out the functionality. Sound good?

sgtm!

**Comment by davidtheclark** [comment]
Just publish 0.3.0 which should provide support for external plugin rules. Unless somebody else wants to tackle it, I can try creating this rule as a separate repo within the next day or two.

**Comment by davidtheclark** [comment]
What does everybody think of this: https://github.com/davidtheclark/stylelint-statement-max-nesting-depth ??

I won't publish it until the nest release of stylelint (when `ruleTester` is exposed in `utils`), but please take a look now, both 
- to see if it fulfills your expectations for this specific rule, and 
- to see if it fulfills your expectations for a stylelint custom rule — since this is the prototype.

**Comment by jeddy3** [comment]
Nice! :)

> to see if it fulfills your expectations for this specific rule

Yeap, looks great to me. @ahmedelgabri Does it fulfil your needs?

> to see if it fulfills your expectations for a stylelint custom rule

Also looks great! I had one query though: I know it was mentioned elsewhere that the configs (when using plugins) will need to be JS, but I'm wondering if it'd be feasible to handle the requiring of plugins within `stylelint` itself. Similar to how `eslint` [does it](http://eslint.org/docs/user-guide/configuring#configuring-plugins):

```js
{
 plugins: [
    'statement-max-nesting-depth',
  ],
  rules: {
    // ...
    'statement-max-nesting-depth': [2, 1], // max nesting depth of 1
    // ...
  },
}
```

Or would you rather keep the requiring of plugins explicit?

**Comment by jeddy3** [comment]
This is probably tied into whether we also provide a similar mechanism hen the user wishes to extend, their own or a one of our, configs. Perhaps via an `extend` key in the config.

Thoughts?

**Comment by ahmedelgabri** [comment]
Good job @davidtheclark!

I had a quick look at it @jeddy3, will plan to test it properly in the weekend.

**Comment by davidtheclark** [comment]
@jeddy3: I prefer using straight JavaScript if this is going to be a JS file. Which means requiring the plugins this way. I do think there's a chance that depending on the stack and knowledge of some CSS authors, there might be users who would prefer to have a configuration file like `.stylelint` that is YAML or JSON. If we get that request (or somebody wants to build it in), we can do as you say.

**Comment by jeddy3** [comment]
> If we get that request (or somebody wants to build it in), we can do as you say.

Sounds groovy to me :)

One last thing before we close this issue... should we put a link to plugin somewhere? Perhaps in the README in the "Plugin Rules" section, or perhaps a new md file called "plugins" (like we have for rules) linked to from the "Plugin Rules" section of the README, or neither?

I'm leaning towards a new md file called "plugins.md", but I'm easy :)

**Comment by davidtheclark** [comment]
Yes, I think this is a good idea. I'm happy with `plugins.md`.

**Comment by davidtheclark** [comment]
Version 0.1.0 is published.

**Comment by jeddy3** [comment]
> Version 0.1.0 is published

Fantastic!

**Comment by ahmedelgabri** [comment]
@davidtheclark :+1: I made a small demo and it works pretty good, thanks!

**Comment by ahmedelgabri** [comment]
One minor issue is, it reports `atRules` too as nesting errors, for example.

```css
.foo {
  color: red;

  > .bar {
    background-color: yellow;

    @media screen and (min-width: 7000px) {
      color: blue;
    }
  }
}
```

This returns an error for `@media` but it shouldn't because this doesn't affect the output of the CSS nesting. The end result will still be `.foo > .bar`.

**Comment by davidtheclark** [comment]
Personally, I would want to include the at rule because it hurts readability just as much as a nested rule. So maybe we should add an option, whether user wants to ignore at rules?

**Comment by ahmedelgabri** [comment]
+1 for the option.

**Comment by davidtheclark** [comment]
Done in v0.3.0.
